docs/ParameterOptimization.md

```diff
@@ -0,0 +1,528 @@
+
+```swift
+struct MyMLModel {
+    // These are parameters.
+    var weight: Tensor<Float>
+    var bias: Tensor<Float>
+
+    // Need to distinguish from non-parameter stored properties.
+    let useBias: Bool
+    var previousWeight: Tensor<Float>
+}
+```
+
+2. There must exist some mechanism to update all parameters of a model given their gradients.
+
+The ability to jointly iterate over parameters and gradients is crucial for writing simple, generic code that works with all models. Without this ability to perform "generic parameter update", users must duplicate code for each parameter, with no potential for generalization:
+
+```swift
+// w1, w2, b1, b2: Tensor<Float>
+w1 -= learningRate * dw1
+w2 -= learningRate * dw2
+b1 -= learningRate * db1
+b2 -= learningRate * db2
+...
+```
+
+### Existing approaches
+
+Dynamic machine learning frameworks like PyTorch allow ad-hoc model parameter registration and mutation. The base class `Module` defines a custom `__setattr__` function, which internally calls `register_parameter`. `Module` also defines a `parameters()` function which returns all registered parameters.
+
+```python
+class Linear(Module):
+  def __init__(self, in_features, out_features, bias=True):
+    super(Linear, self).__init__()
+    self.in_features = in_features
+    self.out_features = out_features
+    # Following line calls `__setattr__`, which internally calls
+    # `register_parameter`.
+    self.weight = Parameter(torch.Tensor(out_features, in_features))
+    if bias:
+      # Following line internally calls `register_parameter`.
+      self.bias = Parameter(torch.Tensor(out_features))
+    else:
+      self.register_parameter('bias', None)
+    self.reset_parameters()
+
+# Optimizers are general and work with the parameters of any model.
+optimizer = optim.SGD(model.parameters(), lr=0.01)
+
+# The optimizer `step` function loops over parameters/parameter groups.
+# Pseudocode:
+class SGD(Optimizer):
+  def step(self):
+    for param in parameters:
+      ...
+```
+
+However, in Swift, this is difficult. Constructing a collection of parameters is difficult because parameters aren't required to have the same type: for example, a model may have parameters with types `Float`, `Tensor<Float>`, and `Tensor<Double>`. To represent parameters as a collection, advanced type-erasure is needed to generalize all parameter types. It's also not clear how parameter update would work (at compile-time, how is it possible to identify the correct concrete `+` function for a type-erased parameter value?).
+
+Additionally, while we want to enable code that achieves the following:
+```swift
+for (inout θ, dθ) in zip(parameters, gradients) {
+    θ -= learningRate * dθ
+}
+```
+
+We don't want to actually lower the for-loop or zip operation to TensorFlow (lowering wouldn't be straightforward or and lowered representation wouldn't be efficient). Instead, we want to fully unroll the loop into individual straight-line statements:
```

Comment:
> We don't want to actually lower the for-loop or zip operation to TensorFlow (lowering wouldn't be straightforward or and lowered representation wouldn't be efficient). Instead, we want to fully unroll the loop into individual straight-line statements:

Note:  this efficiency concern was addressed later in the old version of this doc, but not in this updated version.